imap_processing/ialirt/l0/decom_ialirt.py

```diff
@@ -0,0 +1,142 @@
+                    error_message = f"Error on line {line_number}: " \
+                                    f"Length of binary_values (" \
+                                    f"{len(binary_values)}) does not equal " \
+                                    f"{PacketLength.EXPECTED_LENGTH.value}."
+                    logger.error(error_message)
+                    raise ValueError(error_message)
+
+                packet_generator = packet_parser.generator(
+                    bitstring.ConstBitStream(bin=binary_values))
+
+                for packet in packet_generator:
+                    packets.append(packet)
+
+    return packets
+
+
+def generate_xarray(packet_file: str, xtce: str):
+    """
+    Generate xarray from unpacked data.
+
+    Parameters
+    ----------
+    packet_file : str
+        Path to the CCSDS data packet file.
+    xtce : str
+        Path to the XTCE packet definition file.
+
+    Returns
+    -------
+    xr.Dataset
+        A dataset containing the decoded data fields with 'time' as the coordinating
+        dimension.
+    """
+    try:
+        packets = decom_packets(packet_file, xtce)
+    except Exception as e:
+        logger.error(f"Error during packet decomposition: {str(e)}")
+        return
+
+    if not packets:
+        logger.warning(f"No packets found in {packet_file}.")
+        return
+
+    logger.info(f"Decomposed {len(packets)} packets from {packet_file}.")
+
+    # List of instruments and their corresponding MET keys
+    instruments = ['SC', 'HIT', 'MAG', 'COD_LO', 'COD_HI', 'SWE', 'SWAPI']
+    instrument_coords = ['SC_SCLK_SEC', 'HIT_SC_TICK', 'MAG_ACQ', 'COD_LO_ACQ',
+                         'COD_HI_ACQ', 'SWE_ACQ_SEC', 'SWAPI_ACQ']
+
+    # Create a dictionary mapping each instrument to its time-dimension key
+    time_keys = dict(zip(instruments, instrument_coords))
+
+    # Initialize storage dictionary
+    data_storage = {inst: {} for inst in instruments}
+
+    for packet in packets:
+        for key, value in packet.data.items():
+            for inst in instruments:
+                if key.startswith(inst):
+                    if key not in data_storage[inst]:
+                        data_storage[inst][key] = []
+                    data_storage[inst][key].append(value.derived_value)
+                    break
+            else:
+                logger.warning(f"Unexpected key '{key}' found in packet data.")
+
+    logger.info("Generating datasets for each instrument.")
+
+    # Generate xarray dataset for each instrument and spacecraft
+    datasets = {}
+    for inst in instruments:
+        dataset_dict = {key: (time_keys[inst], value)
+                        for key, value in data_storage[inst].items() if
+                        key != time_keys[inst]}
+        datasets[inst] = xr.Dataset(dataset_dict, coords={
+            time_keys[inst]: data_storage[inst][time_keys[inst]]})
+
+    logger.info(f"Generated datasets for {len(datasets)} instruments.")
```

Comment:
I think this will be the same regardless of how many you actually processed because you define the datasets[inst] based on all your initial instruments you expect. Suggest moving to debug or removing.